### ISSUE TITLE
EVP_CIPHER_CTX_get_app_data/EVP_CIPHER_CTX_set_app_data is not documented

### ISSUE THREAD

**Issue by beldmit** [openssl/openssl#9788]
These functions are only listed in the header of `doc/man3/EVP_EncryptInit.pod` but not documented at all.

I have a particular interest to find out what's the difference between them and EVP_CIPHER_CTX_get_cipher_data/EVP_CIPHER_CTX_set_cipher_data and do we really need both?

**Comment by levitte** [comment]
So...  *implementations* of algorithms may need their own private data, right?  However, the way it's been made is that the EVP API is responsible for actual allocation (using the implementation's `data_size`).  So when the implementation wants to get to its own private data, it has to ask the EVP API where it is, hence `EVP_CIPHER_CTX_get_cipher_data()`.

`EVP_CIPHER_CTX_set_cipher_data()` is a way for the implementation to allocate the private data on its own and *tell* the EVP API where that is.  This is a fairly new addition, and only used in one place in OpenSSL proper, in the dasync engine.  I haven't read the code carefully enough, but I believe it's to be able to switch context between different asynchronous runs.  I'm sure @mattcaswell can fill you in on that.

Either way, algorithm implementations, and especially those in engines need at least one of those functions.

This is not a bug.

Note that with providers, the ballgame is different.  With those, the implementation is fully responsible for its private data.  The EVP API will just receive a pointer to it and pass that along to whatever provided functions that need it.

**Comment by beldmit** [comment]
I agree this is not a bug (I should remove the label myself) and we need these functions, but I do not fill the difference between *app_data and *cipher_data. 

So do we really need both and are these pairs more or less interchangeable?

**Comment by levitte** [comment]
Ah, no, `app_data` is a completely different thing...  that field is exclusively for calling applications that may want to attach its own data to an EVP context object.  Has nothing to do with the algorithm implementations.

**Comment by levitte** [comment]
Of course, one might argue that `app_data` is unnecssary...  that if it somehow wants to pass some undefined out-of-band data to an algorithm implementation, it should do so with control (or parameters in the provider case), and that otherwise, the application should have its own wrapping structure that have the EVP context object as a field.

I frankly don't believe `app_data` is used very much...

**Comment by beldmit** [comment]
Anyway, there is a lack of documentation for these functions. Do we have a special label for it?

**Comment by levitte** [comment]
Good idea to document these.  PR welcome!

**Comment by nhorman** [comment]
@beldmit it seems @levitte asked you for a patch here, did you ever produce one?

**Comment by beldmit** [comment]
No :(
I didn't find any examples of using this API

**Comment by nhorman** [comment]
ok, then i'm not sure this is going to get planned, but the task seems clear, to document 
EVP_CIPHER_CTX_get_app_data/EVP_CIPHER_CTX_set_app_data 

Marking as a community issue

**Comment by t8m** [comment]
This is super-easy documentation addition, marking as good first issue.
